dktest.go

```diff
@@ -167,42 +168,57 @@ func waitContainerReady(ctx context.Context, lgr logger, c ContainerInfo,
 
 // Run runs the given test function once the specified Docker image is running in a container
 func Run(t *testing.T, imgName string, opts Options, testFunc func(*testing.T, ContainerInfo)) {
+	err := RunContext(context.Background(), t, imgName, opts, func(containerInfo ContainerInfo) error {
+		testFunc(t, containerInfo)
+		return nil
+	})
+	if err != nil {
+		t.Fatal("Failed to", err)
+	}
+}
+
+// RunContext is similar to Run, but takes a parent context and returns an error and doesn't rely on a testing.T.
+func RunContext(ctx context.Context, logger Logger, imgName string, opts Options, testFunc func(ContainerInfo) error) (retErr error) {
 	dc, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion("1.41"))
 	if err != nil {
-		t.Fatal("Failed to get Docker client:", err)
+		return fmt.Errorf("get Docker client: %w", err)
 	}
 	defer func() {
-		if err := dc.Close(); err != nil {
-			t.Error("Failed to close Docker client:", err)
+		if err := dc.Close(); err != nil && retErr == nil {
+			retErr = fmt.Errorf("close Docker client: %w", err)
 		}
 	}()
 
 	opts.init()
-	pullCtx, pullTimeoutCancelFunc := context.WithTimeout(context.Background(), opts.PullTimeout)
+	pullCtx, pullTimeoutCancelFunc := context.WithTimeout(ctx, opts.PullTimeout)
 	defer pullTimeoutCancelFunc()
 
-	if err := pullImage(pullCtx, t, dc, imgName, opts.Platform); err != nil {
-		t.Fatal("Failed to pull image:", imgName, "error:", err)
+	if err := pullImage(pullCtx, logger, dc, imgName, opts.Platform); err != nil {
+		return fmt.Errorf("pull image: %v error: %w", imgName, err)
 	}
 
-	func() {
-		runCtx, runTimeoutCancelFunc := context.WithTimeout(context.Background(), opts.Timeout)
+	{
```

Comment:
I removed the `func()` since any defer inside of the func will still behave the same since the func was the last call in the parent function. This allows us to return errors

Reply:
Updated. `retErr` is still used to update the returned error in case the docker client fails to close, but isn't used elsewhere.